api/v1beta1/clusterclass_types.go

```diff
@@ -518,7 +521,125 @@ type JSONSchemaProps struct {
 	// NOTE: Can be set for all types.
 	// +optional
 	Default *apiextensionsv1.JSON `json:"default,omitempty"`
-}
+
+	// XValidations describes a list of validation rules written in the CEL expression language.
+	// +optional
+	XValidations ValidationRules `json:"x-kubernetes-validations,omitempty"`
+}
+
+// ValidationRules describes a list of validation rules written in the CEL expression language.
+// +listType=map
+// +listMapKey=rule
+type ValidationRules []ValidationRule
```

Comment:
Do we want the rule to be also the key in this case? Any downside taht we might see in taht?

Reply:
This is what is used upstream so I think this should be OK. https://github.com/kubernetes/apiextensions-apiserver/blob/v0.29.1/pkg/apis/apiextensions/types_jsonschema.go#L151 Note the different positioning of the build markers due to the way that controller-gen works compared to kubernetes core tooling.